# ***Alexanda Apostolopoulou***
# ***1115201700005***
# *Project 4 in AI (II)* 

# 1st Model of Sentence Embeddings

Import some useful libraries

In [ ]:
import pandas as pd
import numpy as np
import pickle
import glob
import json  
from pandas.io.json import json_normalize  
from nltk.tokenize import sent_tokenize
import nltk
!pip install spicy
import scipy.spatial
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

####***Note:*** I implement both questions 1 and 2 in the same notebook

## ***Step 1:*** Download and Preprocess data

Here I download the initial dataset of COVID-19 Open Research Dataset (CORD-19). Since the dataset is too big, I couldn't upload it to my Drive as in previous Projects, because it's take several hours and it needed to allocate much of my memory. So I decide to download every dataset I would like to experiment with. As you can see I experimented with some different datasets, but I didn't see any improvement (I will refer to this later). The code above is in comments, because after I read the useful data I want from specific files, I saved them to my drive, so it doesn't have to do all this preprocess every time from scratch. 

Download the dataset

In [ ]:
# !wget https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/historical_releases/cord-19_2020-03-13.tar.gz
# !wget https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/historical_releases/cord-19_2020-03-20.tar.gz
# !wget https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/historical_releases/cord-19_2020-05-28.tar.gz

Extract the file

In [ ]:
# !tar -xf cord-19_2020-03-13.tar.gz
# !tar -xf cord-19_2020-03-20.tar.gz

Go to the specific folder

In [ ]:
# %cd /content/2020-03-13
# %cd /content/2020-03-20

Extract the desirable folder

In [ ]:
# !tar -xf comm_use_subset.tar.gz

####Here I take only the title, the abstract and the body texts of each paper and I save them to list of lists. 

####I chose only this three parts of information, because I believe that the sections of paper id, authors, publish_time, url, source, license etc, are irrelevant with the answer I was looking for and they only add more complexity with no reason. 

####In order to understand the structure of my data, the element data[0] is refered to the first article, data[0][0] is refered to the title of the first aricle, data[0][1] to the abstract of the first article and element data[0][2] to the whole body text content of the first paper.

In [ ]:
# data = []

# files = glob.glob('comm_use_subset/*', recursive=True)

# for single_file in files:
#   with open(single_file, 'r') as f:
#     d = json.load(f) 

#     df = pd.json_normalize(d)

#     title = df['metadata.title'].item()

#     if df['abstract'].item() == []:
#       abstract = df['abstract'].item()
#     else:
#       abstract_data = pd.json_normalize(data = d, record_path ='abstract') 
#       if (abstract_data['text'].shape[0] > 1):
#         abstract = abstract_data['text'].str.cat(sep='. ')
#       else:
#         abstract = abstract_data['text'].item()

#     text_data = pd.json_normalize(data = d, record_path ='body_text') 

#     if (text_data['text'].shape[0] > 1):
#       body_text = text_data['text'].str.cat(sep='. ')
#     else:
#       body_text = text_data['text'].item()

#     data.append([title,abstract,body_list])

In [ ]:
# print(len(data))

In [ ]:
# print(data[0])
# print('\n')
# print(data[0][0])
# print('\n')
# print(data[0][1])
# print('\n')
# print(data[0][2])
# print(len(data[0][2]))

Save the previous data locally and then in my Google Drive

In [ ]:
# import os
# cwd = os.getcwd()
# print(cwd)

In [ ]:
# with open('listfile.data', 'wb') as filehandle:
#     # store the data as binary data stream
#     pickle.dump(data, filehandle)

In [ ]:
# with open('listfile.data', 'rb') as filehandle:
#     # read the data as binary data stream
#     articles = pickle.load(filehandle)

In [ ]:
# print(len(articles))
# print(articles[0])
# print('\n')
# print(articles[0][0])
# print('\n')
# print(articles[0][1])
# print('\n')
# print(articles[0][2])

In [ ]:
# import shutil
# shutil.move('/content/listfile_3.data', '/content/drive/MyDrive/listfile_3.data')
# # shutil.move('/content/listfile_2.data', '/content/drive/MyDrive/listfile_2.data')

## ***Step 2:*** Load my preprocessed data from my Drive

In [ ]:
with open('/content/drive/MyDrive/listfile.data', 'rb') as filehandle:
    # read the data as binary data stream
    articles = pickle.load(filehandle)

Check if everything is ok

In [ ]:
# Print the number of articles
print(len(articles))
# Print the first article
print(articles[0])
print('\n')
# Print the first title 
print(articles[0][0])
print('\n')
# Print the first abstract 
print(articles[0][1])
print('\n')
# Print the first body text 
print(articles[0][2])

9000
['ethical and Philosophical considerations for Gain-of-Function Policy: the importance of Alternate experiments The Department of Health and Human Services Framework for Guiding Funding Decisions about Proposed Research Involving Enhanced Potential Pandemic Pathogens', 'PPPs) contains a series of principles for governing the funding and conduct of gain-offunction (GOF) research resulting in the creation of PPPs. In this article, I address one of these principles, governing the replacement of GOF research with alternate experiments. I argue that the principle fails to address the way that different experiments can promote the same values as those promoted by GOF research resulting in PPPs. I then address some objections to this claim, and provide policy recommendations moving forward.', 'Concerns over the accidental or deliberate release of novel pathogens has prompted a debate about the conduct or funding of biological research in the name of human health and security. Of particul

## ***Step 3:*** Tokenize sentences

Here I select to tokenize only the abstract and the body text of each paper, since I wanted to search for the answer in these parts and not in the title. I didn't believe that the answer might be in the title, so I wanted to save more time and memory. Moreover, the titles used to be only one sentence in generall, so it doesn't make much sence to tokenize only one sentence.

In [ ]:
for article in range(len(articles)):
  # Tokenize title
  # articles[article][0] = sent_tokenize(articles[article][0])

  # Tokenize abstract
  if articles[article][1] != []:
    articles[article][1] = sent_tokenize(articles[article][1])
  
  # Tokenize body text
  articles[article][2] = sent_tokenize(articles[article][2])

In [ ]:
print(articles[0][1])

['PPPs) contains a series of principles for governing the funding and conduct of gain-offunction (GOF) research resulting in the creation of PPPs.', 'In this article, I address one of these principles, governing the replacement of GOF research with alternate experiments.', 'I argue that the principle fails to address the way that different experiments can promote the same values as those promoted by GOF research resulting in PPPs.', 'I then address some objections to this claim, and provide policy recommendations moving forward.']


## ***Step 4:*** Download the Sentence Transformers library

In [ ]:
%%capture
!pip install -U sentence-transformers

## ***Step 5:*** Experiment with models of Sentence Transformers

According to the paper ***Sentence-BERT: Sentence Embeddings using Siamese BERT-Networks***, asking BERT to compare sentences is possible but too slow for real-time applications. However, we could pre-compute representations (embeddings) for each sequence in our dataset separate from all other ones. These embeddings could in a second step then be used to measure, for example, similarity using the cosine similarity function which wouldn’t require us to ask BERT to perform this task. SBERT is a so-called twin network which allows it to process two sentences in the same way, simultaneously. SentenceBERT introduces pooling to the token embeddings generated by BERT in order for creating a fixed size sentence embedding. 

From this paper I see that SBERT has better results from any other word embeddings, such as Bert and Glove Embeddings. For this reason I decide to experiment only with diferrent SBERT models.

Here, as you can see, I tried the SBERT models nli-roberta-base and stsb-roberta-base. The large nli roberta took several hours and it never ended (because of crashed session). These two models hadn't many diferrence at the given answers, so I decided to keep only the one of them.

In [ ]:
import sentence_transformers
from sentence_transformers import SentenceTransformer,util
model_ = SentenceTransformer('nli-roberta-base')
# model_ = SentenceTransformer('nli-roberta-large')
# model_ = SentenceTransformer('stsb-roberta-base')

100%|██████████| 460M/460M [00:06<00:00, 74.5MB/s]


Here I took a subset of the whole dataset of 9000 papers, to try different methods, because it takes many hours to run for all the dataset. In the beggining, I experimented with very small datasets until I found the suitable method.

So when I find the appropriate strategy, I run all the notebook for the whole dataset of 9000 articles.

In [ ]:
import copy

mini_articles = articles[:9000]

embeddings = copy.deepcopy(mini_articles)

## ***Step 6:*** Use GPU

In [ ]:
import torch
device = torch.device('cuda:0' if torch.cuda.is_available()
                      else 'cpu')
model_.to(device)

SentenceTransformer(
  (0): Transformer(
    (auto_model): RobertaModel(
      (embeddings): RobertaEmbeddings(
        (word_embeddings): Embedding(50265, 768, padding_idx=1)
        (position_embeddings): Embedding(514, 768, padding_idx=1)
        (token_type_embeddings): Embedding(1, 768)
        (LayerNorm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (dropout): Dropout(p=0.1, inplace=False)
      )
      (encoder): RobertaEncoder(
        (layer): ModuleList(
          (0): RobertaLayer(
            (attention): RobertaAttention(
              (self): RobertaSelfAttention(
                (query): Linear(in_features=768, out_features=768, bias=True)
                (key): Linear(in_features=768, out_features=768, bias=True)
                (value): Linear(in_features=768, out_features=768, bias=True)
                (dropout): Dropout(p=0.1, inplace=False)
              )
              (output): RobertaSelfOutput(
                (dense): Linear(in_features=768, 

## ***Step 7:*** Encode Sentences 

As I said before, I searched for the answer only in abstract and body texts of each paper. For this reason, I concatenate the abstract and the body texts of each paper, so I can encoded them all together. Futhermore, after I encoded to tensors, I kept the original list of text abstarcts and body texts to take the answer in the end.

I calculate the time it takes to compute the embeddings and as you can see it takes a lot of time to find the embeddings to the whole dataset. It is a very time consuming procedure.

In [ ]:
import time
start_time = time.time()

In [ ]:
origin_list_of_articles = []
list_of_articles = []

for article in embeddings:
  article_list = []
  
  # Put both abstract and body texts of each article together
  if article[1] != []:
    for abs_text in article[1]:
      article_list.append(abs_text)

  for body_text in article[2]:
    article_list.append(body_text)

  origin_list_of_articles.append(article_list)
  # Encode all the article 
  list_of_articles.append(model_.encode(article_list,convert_to_tensor=True))

In [ ]:
end_time = time.time()
print(end_time-start_time)

3429.3172903060913


In [ ]:
# print(list_of_articles[0])
# print(len(list_of_articles[0]))
# print(origin_list_of_articles[0])
# print(origin_list_of_articles[0][0])
# print(mini_articles[0][2])
# print(len(list_of_articles[0][0]))
# print(len(list_of_articles))

## ***Step 8:*** Ask model for answers

Here I tried to find the most similar answer of each query, using the measure of ***cosine similarity***. 

***Cosine Similarity*** calculates similarity by measuring the cosine of angle between two vectors. The cosine similarity is advantageous because even if the two similar documents are far apart by the Euclidean distance (due to the size of the document), chances are they may still be oriented closer together. The smaller the angle, higher the cosine similarity.


For the search I found out (after I implement the semantic search by myself :) ), that there is already implemented in the library of sentence_transformers.util.

***semantic_search*** performs a cosine similarity search between a list of query embeddings and a list of corpus embeddings:
 
- Returns a sorted list with decreasing cosine similarity scores. Entries are dictionaries with the keys ‘corpus_id’ and ‘score’.
- I used the parameter top_k  to retrieve ***top 5*** matching entries.

In other words, I took each query and I compared it with each sententce of my data, I save the 5 best scores, the indexes of articles and the indexes of the sentences in a list. 

Finally for each query, I sorted the final list of results to take the 4 most similar sentences in the list. 

In [ ]:
start_time = time.time()

In [ ]:
# Query sentences:
queries = ['What are the coronoviruses?', 
           'What was discovered in Wuhuan in December 2019?', 
           'What is Coronovirus Disease 2019?', 
           'What is COVID-19?', 'What is caused by SARS-COV2?', 
           'How is COVID-19 spread?', 
           'Where was COVID-19 discovered?', 
           'How does coronavirus spread?',
           "Can the use of masks prevent SARS?",
           "What is the Persistence of virus on surfaces of different materials (e.g. copper, stainless steel, plastic)?",
           "What ages have higher mortality rate in covid-19?",
           "What are the protective measures of covid-19?"]

#Encode the queries
query_embeddings = model_.encode(queries, convert_to_tensor=True)

# Find the closest sentence of the corpus for each query sentence based on cosine similarity
for query, query_embedding in zip(queries, query_embeddings):
 
  body_text_distances = []

  for idx_of_article,article in enumerate(list_of_articles):
    cos_scores = sentence_transformers.util.semantic_search(query_embedding, article, top_k=5)

    for elem in cos_scores[0]:
      body_text_distances.append((idx_of_article,elem['corpus_id'],elem['score']))

  results = sorted(body_text_distances, key=lambda x: x[2], reverse=True)

  print("\n=========================================================\n")
  print("Query:", query)
  print("\nThe most similar sentences in corpus:\n")

  print("1. ",origin_list_of_articles[results[0][0]][results[0][1]], "(Score: %.4f)" % (results[0][2]))  
  print("From title: ", mini_articles[results[0][0]][0])

  print('\n')
  print("2. ",origin_list_of_articles[results[1][0]][results[1][1]], "(Score: %.4f)" % (results[1][2]))  
  print("From title: ", mini_articles[results[1][0]][0])

  print('\n')
  print("3. ",origin_list_of_articles[results[2][0]][results[2][1]], "(Score: %.4f)" % (results[2][2]))  
  print("From title: ", mini_articles[results[2][0]][0])

  print('\n')
  print("4. ",origin_list_of_articles[results[3][0]][results[3][1]], "(Score: %.4f)" % (results[3][2]))  
  print("From title: ", mini_articles[results[3][0]][0])



Query: What are the coronoviruses?

The most similar sentences in corpus:

1.  Coronaviruses. (Score: 0.8747)
From title:  DV-Curve Representation of Protein Sequences and Its Application


2.  The Coronaviridae Family. (Score: 0.8689)
From title:  A Genome-Wide Analysis of RNA Pseudoknots That Stimulate Efficient −1 Ribosomal Frameshifting or Readthrough in Animal Viruses


3.  Coccidioides spp. (Score: 0.8634)
From title:  Clinical Practice and Cases


4.  2 | Phylogeny of coronaviruses. (Score: 0.8594)
From title:  Consensus statement The species Severe acute respiratory syndrome- related coronavirus: classifying 2019-nCoV and naming it SARS-CoV-2 Coronaviridae Study Group of the International Committee on Taxonomy of Viruses*


Query: What was discovered in Wuhuan in December 2019?

The most similar sentences in corpus:

1.  In December 2019 (Score: 0.9152)
From title:  Clinical Medicine Characteristics of and Public Health Responses to the Coronavirus Disease 2019 Outbreak in Ch

In [ ]:
end_time = time.time()
print(end_time-start_time)

71.13007307052612


## ***Step 9:*** Summaries

- As you can see the results isn't so satisfactory and I waited one hour to take these results. I thought that the Brute Force Method, to compare the queries with each sentence of the paper, would give me the right answer, but unfortunately this wasn't happened. 

- In the beginning, I print only the first most similar sentence and the results were heartbreaking. It was funny (or nor) to make a question of "What was discovered in Wuhuan in December 2019?" and the given answer was "In December 2019". The score of similarity is very good (0.9152) and we can understand the reason. The results aren't irrational, but there also aren't and the right answer. However if we look more carefully we will see that the title of the answer isn't so bad: "Clinical Medicine Characteristics of and Public Health Responses to the Coronavirus Disease 2019 Outbreak in China".

- For this reason, I decided to show the top 4 similar sentences, to give a chance to see if there were better results with less score, but more relevant with the meaning of the question. As we see, in some questions the results seems to be better like "SARS-2 started in Wuhan in Hubei Province no later than early December 2019." or "Novel Coronavirus (2019-nCoV) is an emerging pathogen that was first identified in Wuhan, China in late December 2019." which is the right answer to the previous question.

- However, there are still questions that the results remain unsatisfied, such as "What is COVID-19?".

- It is very logic to me to have this answers as the best, because for a question that includes the word of COVID-19, the cosine similarity will perform better in documents that have the number 19 inside them. For example the answers of "(IM19)" or "(DOCX) S19 Information" or "1.19)." or "19 )." in the question "What is COVID-19?", isn't so unexpected. Futhermore, these results are because of the sentence tokenize that seperates the sentences where there is a period ".". Maybe if I show more similar answers, there would be a right answer eventually. So this was a chance for me to thing something better than only cosine similarity.

- As I was tried to load subsequently datasets, I didn't see any different in my answers, but only in time, because they have more articles.

- I have to mention that I added some of my own questions to see the results like "What ages have higher mortality rate in covid-19?" , whose almost all of the answers are right. Another questions that I give are "Can the use of masks prevent SARS?" or "What are the protective measures of covid-19?" or "What is the Persistence of virus on surfaces of different materials (e.g., copper, stainless steel, plastic)?" that they didn't go very well.

- As for the pretrained nli-roberta-base that I chose, it was slightly slower than stsb-roberta-base' (~1 second slower for each article), but because I didn't saw any basic difference in results, I kept the nli-roberta-base pretrained model, as according to the paper it performs better results.

- I have to refer that I tried to take with cosine similarity, the top most similar titles, and then for these 100 articles I took the top 50 most similar abstracts and finally the 20 most similar text bodies. However the results was even worse and for this reason I decided to not present them to you (it is the code of comments below). Besides the search of the answer didn't take so much time (only ~1 minute), to have a reason to reduce data. The time consuming procedure is to encode the sentences of each data.

- So I decided to not encode all the dataset for the answer to my second model.

In [ ]:
# # Query sentences:
# queries = ['What are the coronoviruses?', 
#            'What was discovered in Wuhuan in December 2019?', 
#            'What is Coronovirus Disease 2019?', 
#            'What is COVID-19?', 'What is caused by SARS-COV2?', 
#            'How is COVID-19 spread?', 
#            'Where was COVID-19 discovered?', 
#            'How does coronavirus spread?']

# query_embeddings = model.encode(queries)

# # Find the closest sentence of the corpus for each query sentence based on cosine similarity
# for query, query_embedding in zip(queries, query_embeddings):
 
#   title_distances = []
#   for article in range(len(mini_articles)):
#     title_distances.append(scipy.spatial.distance.cdist([query_embedding], [embeddings[article][0]], "cosine")[0])

#   results = zip(range(len(title_distances)), title_distances)
#   results = sorted(results, key=lambda x: x[1])

#   top = 100
#   abs_distances = []
#   for idx, distance in results[0:top]:
#     sum = 0
#     num_of_sentences = 0
#     for sentence in range(len(mini_articles[idx][1])):
#       sum += scipy.spatial.distance.cdist([query_embedding], [embeddings[idx][1][sentence]], "cosine")[0]
#       num_of_sentences += 1
#     if num_of_sentences != 0:
#       avg = sum/num_of_sentences
#     abs_distances.append((idx,avg))

#   results = sorted(abs_distances, key=lambda x: x[1])

#   top = 50
#   body_text_distances = []
  
#   for idx, distance in results[0:top]:
#     distances = []
#     num_of_sentence = 0
#     for sentence in range(len(embeddings[idx][2])):
#       distances.append((num_of_sentence, (scipy.spatial.distance.cdist([query_embedding], [embeddings[idx][2][sentence]], "cosine")[0])))
#       num_of_sentence += 1
    
#     results = sorted(distances, key=lambda x: x[1])

#     body_text_distances.append((idx,results[0][0], results[0][1]))

#   results = sorted(body_text_distances, key=lambda x: x[2])

#   print("\n======================\n")
#   print("Query:", query)
#   print("\nThe most similar sentence in corpus:\n")

#   print(origin_list_of_articles[results[0][0]][results[0][1]], "(Score: %.4f)" % (1-results[0][2]))  